### PR TITLE
Refactoring usageClient with Interface

### PR DESCRIPTION
**Pull request by sajithab**

### CODE REVIEW

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+     *
+     * @param providerName Name of the WebApp provider
+     * @param fromDate
+     * @param toDate
+     * @param limit Number of entries to return
+     * @param tenantDomainName
+     * @return a List of AppUsageDTO objects - possibly empty
+     * @throws org.wso2.carbon.appmgt.api.exception.AppUsageQueryServiceClientException if an error occurs while contacting backend services
+     */
+    public List<AppUsageDTO> getUsageByAPIs(String providerName, String fromDate, String toDate,
```
Comment (lakshani) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java]: getUsageByAPIs >> getUsageByApps

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+     * @param fromDate
+     * @param toDate
+     * @return a List of AppResourcePathUsageDTO objects, possibly empty
+     * @throws org.wso2.carbon.appmgt.api.exception.AppUsageQueryServiceClientException on error
+     */
+    public List<AppResourcePathUsageDTO> getAPIUsageByResourcePath(String providerName, String fromDate,
+                                                                   String toDate)
+            throws AppUsageQueryServiceClientException;
+
+    public List<AppPageUsageDTO> getAPIUsageByPage(String providerName, String fromDate, String toDate,
```
Comment (lakshani) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java]: method comments

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+    /**
+     * Get cache hit summary.
+     *
+     * @param providerName Name of the WebApp provider
+     * @param fromDate String
+     * @param toDate   String
+     * @return AppMCacheCountDTO App cache hits stats list.
+     * @throws org.wso2.carbon.appmgt.api.exception.AppUsageQueryServiceClientException
+     */
+    public List<AppMCacheCountDTO> getCacheHitCount(String providerName, String fromDate, String toDate)
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java]: remove

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+            throws AppUsageQueryServiceClientException;
+
+    public List<PerUserAPIUsageDTO> getUsageBySubscribers(String providerName, String apiName,
+                                                          String apiVersion, int limit)
+            throws AppUsageQueryServiceClientException;
+
+    public List<AppVersionUserUsageDTO> getUsageBySubscriber(String subscriberName, String period)
+            throws Exception;
+
+    public List<String> getFirstAccessTime(String providerName, int limit) throws AppUsageQueryServiceClientException;
```
Comment (lakshani) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/AppUsageStatisticsClient.java:248]: method comments

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/dto/AppVersionUsageDTO.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wso2.carbon.appmgt.api.dto;
+
+public class AppVersionUsageDTO {
```
Comment (lakshani) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/dto/AppVersionUsageDTO.java:19]: class comments

File: components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.client/src/main/java/org/wso2/carbon/appmgt/usage/client/internal/AppMUsageClientServiceComponent.java [hunk 6]
```diff
@@ -18,47 +18,89 @@
...
+        BundleContext bundleContext = componentContext.getBundleContext();
+        //Register the default App usage stat client as a OSGi service.
+        bundleContext.registerService(AppUsageStatisticsClient.class.getName(), new AppUsageStatisticsRdbmsClient(), null);
+
+        //Find the app usage statistics client which is proffered in the configuration and set instance of it.
+        appUsageStatisticsClientImplClazz = configuration.getFirstProperty(APIMgtUsagePublisherConstants.APP_STATISTIC_CLIENT_PROVIDER);
+        doRegisterAppUsageStatisticsClient();
+    }
+
+    private void doRegisterAppUsageStatisticsClient() {
```
Comment (sajithab) [inline components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.client/src/main/java/org/wso2/carbon/appmgt/usage/client/internal/AppMUsageClientServiceComponent.java]: move to impl

File: components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.client/src/main/java/org/wso2/carbon/appmgt/usage/client/internal/AppMUsageClientServiceComponent.java [hunk 7]
```diff
@@ -18,47 +18,89 @@
...
         if (log.isDebugEnabled()) {
-            log.debug("WebApp usage client component activated");
+            log.debug("org.wso2.appmgt.usage.client component has been activating");
+        }
+        BundleContext bundleContext = componentContext.getBundleContext();
+        //Register the default App usage stat client as a OSGi service.
+        bundleContext.registerService(AppUsageStatisticsClient.class.getName(), new AppUsageStatisticsRdbmsClient(), null);
+
+        //Find the app usage statistics client which is proffered in the configuration and set instance of it.
+        appUsageStatisticsClientImplClazz = configuration.getFirstProperty(APIMgtUsagePublisherConstants.APP_STATISTIC_CLIENT_PROVIDER);
```
Comment (lakshani) [inline components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.client/src/main/java/org/wso2/carbon/appmgt/usage/client/internal/AppMUsageClientServiceComponent.java]: appUsageStatisticsClientImplClazz >> appUsageStatisticsClientImplClass

File: components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.client/src/main/java/org/wso2/carbon/appmgt/usage/client/pojo/AppResponseTime.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                APIUsageStatisticsClientConstants.CONTEXT)).getText();
+        responseTime = Double.parseDouble(row.getFirstChildWithName(new QName(
+                APIUsageStatisticsClientConstants.SERVICE_TIME)).getText());
+        page = row.getFirstChildWithName(new QName(APIUsageStatisticsClientConstants.REFERER)).getText();
+        pageName = page.split("//")[1].split("/");
+        for(int x = 1;x<pageName.length;x++){
+            referer =referer + "/"+pageName[x] ;
+        }
+
+            /*responseCount = (long) Double.parseDouble(row.getFirstChildWithName(new QName(
```
Comment (lakshani) [inline components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.client/src/main/java/org/wso2/carbon/appmgt/usage/client/pojo/AppResponseTime.java]: remove unwanted comments

File: components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/dto/AppUsageDTO.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.wso2.carbon.appmgt.api.dto;
+
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.api/src/main/java/org/wso2/carbon/appmgt/api/dto/AppUsageDTO.java:18]: Add class comments

File: components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.publisher/src/main/java/org/wso2/carbon/appmgt/usage/publisher/APIMgtUsagePublisherConstants.java [hunk 10]
```diff
@@ -31,11 +31,11 @@
     public static final String API_USAGE_TRACKING = "Analytics.";
     public static final String API_USAGE_ENABLED = API_USAGE_TRACKING + "Enabled";
     public static final String API_USAGE_THRIFT_PORT = API_USAGE_TRACKING + "ThriftPort";
-    public static final String API_USAGE_BAM_SERVER_URL = API_USAGE_TRACKING + "BAMServerURL";
-    public static final String API_USAGE_BAM_SERVER_USER = API_USAGE_TRACKING + "BAMUsername";
-    public static final String API_USAGE_BAM_SERVER_PASSWORD = API_USAGE_TRACKING + "BAMPassword";
+    public static final String API_USAGE_DAS_SERVER_URL = API_USAGE_TRACKING + "DASServerURL";
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.publisher/src/main/java/org/wso2/carbon/appmgt/usage/publisher/APIMgtUsagePublisherConstants.java]: Move these constant definitions into the Constant class

File: components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.publisher/src/main/java/org/wso2/carbon/appmgt/usage/publisher/APPMgtUsageHandler.java [hunk 11]
```diff
@@ -31,25 +31,18 @@
...
 
 public class APPMgtUsageHandler extends AbstractHandler {
-	
-	private static final String URL_SEPERATOR = "/";
-	private static final String API_AUTH_CONTEXT = "__API_AUTH_CONTEXT";
-
-    private static final Log log   = LogFactory.getLog(APPMgtUsageHandler.class);
 
+    private static final String URL_SEPERATOR = "/";
+    private static final String API_AUTH_CONTEXT = "__API_AUTH_CONTEXT";
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.publisher/src/main/java/org/wso2/carbon/appmgt/usage/publisher/APPMgtUsageHandler.java]: Move these constant definitions into the Constant class

File: components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.publisher/src/main/java/org/wso2/carbon/appmgt/usage/publisher/APPMgtUsageHandler.java [hunk 12]
```diff
@@ -71,101 +64,104 @@ public boolean handleRequest(MessageContext mc) {
...
-	}
-    
-    public boolean handleResponse(MessageContext mc) {
+            AuthenticationContext authContext = (AuthenticationContext) messageContext.getProperty(API_AUTH_CONTEXT);
+
+            if (publisher == null) {
+                synchronized (this) {
+                    if (publisher == null) {
+                        try {
+                            log.debug("Instantiating Data Publisher");
```
Comment (tgtshanika) [inline components/org.wso2.carbon.appmgt.usage/org.wso2.carbon.appmgt.usage.publisher/src/main/java/org/wso2/carbon/appmgt/usage/publisher/APPMgtUsageHandler.java:93]: check fo isDebugEnable true/false

**Pull request merged**